doc/partfs.man
.TH "PARTFS" 1 "October 2018" "PartFS version 1.0"

.\"-----------------------------------------------------------------------------

.SH NAME 
.B partfs
\(en Mount a region of \fISOURCE\fR as a different file at \fIMOUNTPOINT\fR.

.\"-----------------------------------------------------------------------------

.SH SYNOPSIS
.B partfs
.RB [ \-h/--help | \-V/--version ]
.LP
.PD 0
.B partfs
[-o \fBoptions\fR]
.I SOURCE
.I MOUNTPOINT
.LP

.B partfs
.I SOURCE
.I MOUNTPOINT
[-o \fBoptions\fR]
.LP
.LP
.B partfs
.I SOURCE
[-o \fBoptions\fR]
.I MOUNTPOINT
.PD

.SS Size/Offset Examples:
.PD 0
.B partfs
.I SOURCE
.I MOUNTPOINT
-o offset=\fINBYTES\fR[,\fBoptions\fR]
.LP

.B partfs
.I SOURCE
.I MOUNTPOINT
-o sizelimit=\fINBYTES\fR[,\fBoptions\fR]
.LP

.B partfs
.I SOURCE
.I MOUNTPOINT
-o offset=\fINBYTES\fR,sizelimit=\fINBYTES\fR[,\fBoptions\fR]
.PD
.LP

.SS Common FUSE Options:
.PD 0

.B partfs
.I SOURCE
.I MOUNTPOINT
-o ro (mounts as read-only)
.LP

.B partfs
.I SOURCE
.I MOUNTPOINT
-o allow_other (allows other users to access mount)
.LP

.B partfs
.I SOURCE
.I MOUNTPOINT
-o nonempty (allow mounting over a non-empty file, or over itself)
.PD
.LP

.SS Unmounting:
\fBfusermount\fR \fB-u\fR \fIMOUNTPOINT\fR

.RS -4
See \fBpartfs --help\fR for a full list of mouting options.
.RE

.\"-----------------------------------------------------------------------------

.SH DESCRIPTION 

PartFS is a FUSE-based tool for accessing a chunk of a SOURCE file as if it
were a separate file at MOUNTPOINT. It's intended for use in creating/modifying
disk image files.
.LP

Say that SOURCE was a 600000-byte file with the following layout:
.LP
.PD 0
.RS 4n
[Region 1] Bytes 1 to 5000: Bootloader region
.LP
[Region 2] Bytes 5001 to 105000: System data
.LP
[Region 3] Bytes 105001 to 205000: System Partition
.LP
[Region 4] Bytes 205001 to 600000: User Data
.RE
.PD
.LP
PartFS would allow you to mount any of these regions and access them as if
they were standalone files. In the example given above, the correct mounting
parameters would be:
.LP
.PD 0
.RS 4n
partfs \fISOURCE\fR \fIMOUNTPOINT\fR -o offset=0,sizelimit=5000
.LP
partfs \fISOURCE\fR \fIMOUNTPOINT\fR -o offset=5000,sizelimit=100000
.LP
partfs \fISOURCE\fR \fIMOUNTPOINT\fR -o offset=105000,sizelimit=100000
.LP
partfs \fISOURCE\fR \fIMOUNTPOINT\fR -o offset=205000,sizelimit=395000
.RE
.PD
.LP
Note that region 1 could also be mounted with no offset because it starts at
byte 0. Similarly, Region 4 could be mounted with no size-limit because it goes
to the end of the file.
.LP

PartFS is logically similar to a loopback-mounted image under Linux. It has two
key differences though. First, it can be done as an unprivileged user. And
second, it doesn't offer direct access to any filesystems stored within the
mounted image. Consider using with fuse2fs or a similar tool for unprivileged
filesystem access.

\fIMOUNTPOINT\fR must be an empty file unless the \fBnonempty\fR option is
supplied, as described under 'General FUSE options' below.

PartFS exposes the full set of FUSE mounting options. These include \fIro\fR for
read-only mounts, \fIallow_other\fR for multi-user mounts, \fInonempty\fR for
self-mounts or mounting over non-empty files, and many more. See
\fBpartfs --help\fR for a complete list.

To unmount a PartFS file, use the \fBfusermount\fR utility.

.\"-----------------------------------------------------------------------------

.SH OPTIONS
.SS General Options
.TP
.B -h, --help
Display the help message and exit. Includes FUSE's help message.

.TP
.B -V, --version
Display the program version-number and exit.

.SS Mount Options
Like many other fileysstem-mount programs, mount options can be specified in
a single argument (-o \fIOPT1,OPT2=val2,OPT3\fR) as a list of multiple
arguments (-o \fIOPT1\fR -o \fIOPT2=val2\fR -o \fIOPT3\fR), or as a mixture of
the two styles.
.LP
.RS -2n
PartFS-specific options:
.RE
.TP
.B -o offset=NBYTES
The number of bytes to skip in \fISOURCE\fR (default: 0). Byte 1 of
\fIMOUNTPOINT\fR will map to byte [\fBoffset\fR + 1] of \fISOURCE\fR, as
described in the \fBsizelimit\fR option below. \fISOURCE\fR's contents below
the offset are never modified.

.TP
.B -o sizelimit=NBYTES
Starting length (and maxiumum length) of \fIMOUNTPOINT\fR in bytes. The
mapped region of \fISOURCE\fR will go from byte [\fBoffset\fR] to byte
[\fBsizelimit\fR-1], including both edges. So a mount with a 1500-byte sizelimit
will and no offset will include bytes 0-1499 (for a total length of 1500).
\fISOURCE\fR's contents above [offset + sizelimit] are never modified.

.RS -2n
General FUSE options:
.RE
.LP
There are a lot of FUSE options available, all of which are provided by FUSE
instead of directly from PartFS. A complete list can be found by running 
\fBpartfs --help\fR. Some of the more interesting options are documented here:

.TP
.B -o ro
Mounted file is read-only, and can't be modified by any user.

.TP
.B -o allow_root
Mounted file can be accessed by root as well as the mounter. This option and
\fBallow_other\fR are mutually exclusive.

.TP
\fB-o allow_other\fR (if enabled on your system)
Mounted file can be accessed by any user. This option and \fBallow_root\fR are
mutually exclusive.

.TP
.B -o nonempty
Allow mounting over a non-empty file. By default, FUSE will only mount to empty
files (in order to avoid hiding data). If supplied, this option will allow you
to mount on-top of any file (even including \fISOURCE\fR) if desired.

.TP
.B -f
Run in the foreground instead of starting a background process and returning.

.\"-----------------------------------------------------------------------------

.SH NOTES

PartFS is a FUSE filesystem, and will support slightly different options and
unmount procedures depending on your system's FUSE installation.

Also note that PartFS is a file-to-file mount, and doesn't give you direct
access to an image's filesystem. To edit a filesystem, a secondary mount (using
\fBfuse2fs\fR or a similar tool) is required.

.SH SEE ALSO
.SS \fRManpages:
.LP
.PD 0
.BR fuse( 4 ),
.LP
.BR fusermount( 1 ),
.LP
.BR fuse2fs( 1 ),
.LP
.BR mount.fuse( 8 ),
.LP
.PD
.SS \fRURL:
.I https://github.com/nrclark/partfs



partfs.c
/*
 *  PartFS: A FUSE filesystem for mounting parts of a file.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.

 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.

 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 *  Copyright 2018, Nicholas Clark */

#define _POSIX_C_SOURCE 200809L
#define FUSE_USE_VERSION 26

#include <errno.h>
#include <fuse.h>
#include <libgen.h>
#include <libintl.h>
#include <limits.h>
#include <locale.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "config.h"
#define _(x) gettext(x)

/*----------------------------------------------------------------------------*/

static char progname[NAME_MAX + 1] = {0};

struct partfs_context {
    int read_only;
    int source_fd;
    mode_t source_mode;
    size_t max_size;
    size_t source_offset;
    size_t current_size;
    struct fuse_args *args;
};

struct partfs_config {
    size_t offset;
    size_t size;
    int read_only;
    int nonempty;
    char source[PATH_MAX + 1];
    char mountpoint[PATH_MAX + 1];
};

#define PARTFS_OPT(t, p, v) {t, offsetof(struct partfs_config, p), v}

enum {
    KEY_VERSION,
};

static struct fuse_opt partfs_opts[] = {
    PARTFS_OPT("offset=%zu", offset, 0),
    PARTFS_OPT("sizelimit=%zu", size, 0),
    PARTFS_OPT("ro", read_only, 1),
    PARTFS_OPT("nonempty", nonempty, 1),

    FUSE_OPT_KEY("-V", KEY_VERSION),
    FUSE_OPT_KEY("--version", KEY_VERSION),
    FUSE_OPT_KEY("-h", 0),
    FUSE_OPT_KEY("--help", 0),
    FUSE_OPT_END
};

/*----------------------------------------------------------------------------*/

static void controlled_exit(struct partfs_context *ctx, int exit_code)
{
    if (ctx != NULL) {
        if (ctx->source_fd >= 0) {
            close(ctx->source_fd);
        }

        if (ctx->args != NULL) {
            fuse_opt_free_args(ctx->args);
        }
    }

    exit(exit_code);
}

static ssize_t read_noeintr(int fildes, void *buf, size_t nbyte)
{
    ssize_t result;

    do {
        result = read(fildes, buf, nbyte);
    } while ((result == -1) && (errno == EINTR));

    return result;
}

static ssize_t write_noeintr(int fildes, const void *buf, size_t nbyte)
{
    ssize_t result;

    do {
        result = write(fildes, buf, nbyte);
    } while ((result == -1) && (errno == EINTR));

    return result;
}

static ssize_t read_count(int filedes, char *buf, size_t nbyte)
{
    size_t total = 0;

    while (total < nbyte) {
        ssize_t result = read_noeintr(filedes, buf, (nbyte - total));

        if (result < 0) {
            return result;
        }

        total += (size_t) result;
    }

    return (ssize_t) nbyte;
}

static ssize_t write_count(int filedes, const char *buf, size_t nbyte)
{
    size_t remaining = nbyte;

    while (remaining != 0) {
        ssize_t result = write_noeintr(filedes, buf, remaining);

        if (result < 0) {
            return result;
        }

        remaining -= (size_t) result;
        buf += result;
    }

    return (ssize_t) nbyte;
}

static inline void safecopy(char *dest, const char *src, unsigned int maxlen)
{
    size_t length = strnlen(src, maxlen - 1);
    memcpy(dest, src, length);
    dest[length] = '\x00';
}

static void exit_help(void)
{
    const struct fuse_operations dummy_ops = {0};
    char help_opt[4] = "-ho";
    char *argv[2] = {progname, help_opt};

    const char *partfs_help =
        _("Mount part of SOURCE as a different file at MOUNTPOINT.\n"
          "\n"
          "Usage: %s SOURCE MOUNTPOINT [options]\n"
          "\n"
          "General options:\n"
          "    -o opt,[opt...]        mount options\n"
          "    -h   --help            print help\n"
          "    -V   --version         print version\n"
          "\n"
          "PartFS options:\n"
          "    -o offset=NBYTES       offset into SOURCE (in bytes)\n"
          "    -o sizelimit=NBYTES    max length of MOUNT (in bytes)");

    fprintf(stderr, partfs_help, progname);
    fprintf(stderr, "\n\n");
    fuse_main(2, argv, &dummy_ops, NULL);
    exit(1);
}

static int partfs_opt_proc(void *data, const char *arg, int key,
                           struct fuse_args *outargs)
{
    struct partfs_config *config = (struct partfs_config *) data;
    const struct fuse_operations dummy_ops = {0};

    switch (key) {
        case KEY_VERSION:
            fprintf(stderr, _("PartFS version: %s"), PACKAGE_VERSION);
            fprintf(stderr, "\n");
            fuse_opt_add_arg(outargs, "--version");
            fuse_main(outargs->argc, outargs->argv, &dummy_ops, NULL);
            fuse_opt_free_args(outargs);
            exit(0);
            break;

        case FUSE_OPT_KEY_NONOPT:
            if (config->source[0] == '\x00') {
                if (arg[0] == '\x00') {
                    fprintf(stderr, "%s: ", progname);
                    fprintf(stderr, "%s\n",
                            _("error: source must not be an empty string."));
                    fuse_opt_free_args(outargs);
                    exit(1);
                }
                safecopy(config->source, arg, sizeof(config->source));
                return 0;
            }

            if (config->mountpoint[0] == '\x00') {
                if (arg[0] == '\x00') {
                    fprintf(stderr, "%s: ", progname);
                    fprintf(stderr, "%s\n",
                            _("error: mount-point must not be an empty string.")
                            );
                    fuse_opt_free_args(outargs);
                    exit(1);
                }
                safecopy(config->mountpoint, arg, sizeof(config->mountpoint));
                return 1;
            }

            fprintf(stderr, "%s: %s [%s]\n", progname, _("error: invalid "
                                                         "additional argument"),
                    arg);

            fuse_opt_free_args(outargs);
            exit(1);
            break;

        default:
            break;
    }
    return 1;
}

static struct partfs_context * partfs_get_context(void)
{
    struct fuse_context *fuse_context = fuse_get_context();
    struct partfs_context *result;

    if (fuse_context == NULL) {
        fprintf(stderr, "%s: ", progname);
        fprintf(stderr, "%s\n", _("error: couldn't retrieve FUSE context."));
        exit(0);
    }

    result = (struct partfs_context *)(fuse_context->private_data);

    if (result == NULL) {
        fprintf(stderr, "%s: ", progname);
        fprintf(stderr, "%s\n", _("error: couldn't retrieve PartFS context."));
        exit(0);
    }

    return result;
}

/*----------------------------------------------------------------------------*/

static int partfs_open(const char *path, struct fuse_file_info *info)
{
    (void) path;
    (void) info;
    struct partfs_context *ctx = partfs_get_context();

    if (ctx->read_only && ((info->flags & O_ACCMODE) != O_RDONLY)) {
        return -EACCES;
    }

    return 0;
}

static int partfs_getattr(const char *path, struct stat *stbuf)
{
    (void) path;
    struct stat source_stat;
    int result;
    struct partfs_context *ctx = partfs_get_context();

    memset(stbuf, 0, sizeof(struct stat));

    stbuf->st_uid = getuid();
    stbuf->st_gid = getgid();

    stbuf->st_mode = S_IFREG | ctx->source_mode;

    if (ctx->read_only) {
        stbuf->st_mode &= ~0222U;
    }

    stbuf->st_nlink = 1;
    stbuf->st_size = (off_t) ctx->current_size;

    result = fstat(ctx->source_fd, &source_stat);

    if (result < 0) {
        return -errno;
    }

    memcpy(&(stbuf->st_atim), &(source_stat.st_atim), sizeof(stbuf->st_atim));
    memcpy(&(stbuf->st_mtim), &(source_stat.st_mtim), sizeof(stbuf->st_mtim));
    memcpy(&(stbuf->st_ctim), &(source_stat.st_ctim), sizeof(stbuf->st_ctim));

    return 0;
}

static int partfs_read(const char *path, char *buf, size_t size,
                       off_t offset, struct fuse_file_info *info)
{
    (void) path;
    off_t lseek_result;
    ssize_t read_result;
    struct partfs_context *ctx = partfs_get_context();

    if (((size_t)offset + size) < (size_t)offset) {
        /* Size + offset overflowed */
        return -EINVAL;
    }

    if (((size_t)offset + size) > ctx->current_size) {
        size = ctx->current_size - (size_t) offset;
    }

    if (size == 0) {
        return 0;
    }

    lseek_result = lseek(ctx->source_fd, offset + (off_t) ctx->source_offset,
                         SEEK_SET);

    if (lseek_result < 0) {
        if (info->direct_io) {
            return (int) lseek_result;
        }
        return -errno;
    }

    if (info->direct_io) {
        return (int) read(ctx->source_fd, buf, size);
    }

    read_result = read_count(ctx->source_fd, buf, size);

    if (read_result < 0) {
        return -errno;
    }

    return (int) read_result;
}

static int partfs_write(const char *path, const char *buf, size_t size,
                        off_t offset, struct fuse_file_info *info)
{
    (void) path;
    off_t lseek_result;
    ssize_t write_result;
    size_t stop_byte;

    struct partfs_context *ctx = partfs_get_context();

    if (((size_t)offset + size) < (size_t)offset) {
        /* Size + offset overflowed */
        return -EINVAL;
    }

    if ((size_t)offset > ctx->max_size) {
        return -EIO;
    }

    if (((size_t)offset == ctx->max_size) && (size != 0)) {
        return -EIO;
    }

    if (((size_t)offset + size) > ctx->max_size) {
        size = ctx->max_size - (size_t) offset;
    }

    stop_byte = (size_t) offset + size;

    if (stop_byte > ctx->current_size) {
        ctx->current_size = stop_byte;
    }

    lseek_result = lseek(ctx->source_fd, offset + (off_t) ctx->source_offset,
                         SEEK_SET);

    if (lseek_result < 0) {
        if (info->direct_io) {
            return (int) lseek_result;
        }
        return -errno;
    }

    if (info->direct_io) {
        return (int) write(ctx->source_fd, buf, size);
    }

    write_result = write_count(ctx->source_fd, buf, size);

    if (write_result < 0) {
        return -errno;
    }

    return (int) write_result;
}

static int partfs_access(const char *path, int amode)
{
    (void) path;
    struct partfs_context *ctx = partfs_get_context();

    if ((amode & W_OK) && ctx->read_only) {
        return -EACCES;
    }

    if (amode & X_OK) {
        return -EACCES;
    }

    return 0;
}

static int partfs_utimens(const char *path, const struct timespec tv[2])
{
    (void) path;
    struct partfs_context *ctx = partfs_get_context();

    int result = futimens(ctx->source_fd, tv);

    if (result < 0) {
        return -errno;
    }
    return 0;
}

static int partfs_truncate(const char *path, off_t length)
{
    (void) path;
    struct partfs_context *ctx = partfs_get_context();

    ctx->current_size = (size_t) length;
    return 0;
}

static int partfs_chown(const char *path, uid_t uid, gid_t gid)
{
    (void) path;
    (void) uid;
    (void) gid;
    return -EPERM;
}

static int partfs_chmod(const char *path, mode_t mode)
{
    (void) path;
    (void) mode;
    return 0;
}

static int partfs_fsync(const char *path, int datasync,
                        struct fuse_file_info *info)
{
    (void) path;
    (void) datasync;
    (void) info;
    struct partfs_context *ctx = partfs_get_context();

    int result = fsync(ctx->source_fd);

    if (result < 0) {
        return -errno;
    }

    return result;
}

/*----------------------------------------------------------------------------*/

static struct fuse_operations partfs_operations = {
    .getattr = partfs_getattr,
    .open = partfs_open,
    .read = partfs_read,
    .write = partfs_write,
    .access = partfs_access,
    .utimens = partfs_utimens,
    .truncate = partfs_truncate,
    .chown = partfs_chown,
    .chmod = partfs_chmod,
    .fsync = partfs_fsync,
};

int main(int argc, char *argv[])
{
    struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
    struct partfs_config config = {.size = (size_t) -1};
    struct partfs_context context = {.source_fd = -1, .args = &args};

    char arg_buffer[sizeof("-ofsname=") + NAME_MAX + 2] = "-ofsname=";
    unsigned int arg_offset = sizeof("-ofsname=") - 1;
    unsigned int arg_maxlen = (unsigned int) sizeof(arg_buffer) - arg_offset;

    struct stat stat_buffer;
    int result;

    setlocale(LC_ALL, "");
#ifdef ENABLE_NLS
    if (getenv("LOCALEDIR")) {
        bindtextdomain(PACKAGE, getenv("LOCALEDIR"));
    } else {
        bindtextdomain(PACKAGE, LOCALEDIR);
    }
    textdomain(PACKAGE);
#endif

    safecopy(progname, basename(argv[0]), sizeof(progname));

    for (int x = 1; x < argc; x++) {
        if ((strcmp(argv[x], "--help") == 0) || (strcmp(argv[x], "-h") == 0)) {
            fuse_opt_free_args(&args);
            exit_help();
        }
    }

    fuse_opt_parse(&args, &config, partfs_opts, partfs_opt_proc);

    if (config.source[0] == '\x00') {
        fprintf(stderr, "%s: %s\n", progname,
                _("error: source not specified."));
        controlled_exit(&context, 1);
    }

    if (config.mountpoint[0] == '\x00') {
        fprintf(stderr, "%s: %s\n", progname,
                _("error: mount-point not specified."));
        controlled_exit(&context, 1);
    }

    result = lstat(config.mountpoint, &stat_buffer);

    if (result != 0) {
        fprintf(stderr, "%s: ", progname);
        fprintf(stderr, _("error: couldn't open mount-point [%s]"),
                config.mountpoint);
        fprintf(stderr, " (%s)\n", strerror(errno));
        controlled_exit(&context, 1);
    }

    if (((stat_buffer.st_size != 0) && (config.nonempty == 0)) ||
        (S_ISREG(stat_buffer.st_mode) == 0)) {
        fprintf(stderr, "%s: %s\n", progname,
                _("error: mount-point is not an empty file."));
        controlled_exit(&context, 1);
    }

    if (config.read_only) {
        context.source_fd = open(config.source, O_RDONLY);
    } else {
        context.source_fd = open(config.source, O_RDWR);
    }

    if (context.source_fd < 0) {
        fprintf(stderr, "%s: ", progname);
        fprintf(stderr, _("error: couldn't open file [%s]"),
                config.source);
        fprintf(stderr, " (%s)\n", strerror(errno));
        controlled_exit(&context, 1);
    }

    result = stat(config.source, &stat_buffer);

    if (result != 0) {
        fprintf(stderr, "%s: ", progname);
        fprintf(stderr, _("error: couldn't stat file [%s]"),
                config.source);
        fprintf(stderr, " (%s)", strerror(errno));
        controlled_exit(&context, 1);
    }

    if (config.size == (size_t) -1) {
        config.size = (size_t) stat_buffer.st_size - config.offset;
    }

    if ((config.offset + config.size) > (size_t) stat_buffer.st_size) {
        fprintf(stderr, "%s: ", progname);
        fprintf(stderr, _("error: requested size or offset extends past the"
                          " end of [%s]"), basename(config.source));
        fprintf(stderr, "\n");
        controlled_exit(&context, 1);
    }

    context.source_mode = stat_buffer.st_mode;
    context.read_only = config.read_only;
    context.max_size = config.size;
    context.current_size = config.size;
    context.source_offset = config.offset;

    fuse_opt_add_arg(&args, "-s");

    if (config.nonempty) {
        fuse_opt_add_arg(&args, "-ononempty");
    }

    safecopy(arg_buffer + arg_offset, config.source, arg_maxlen);
    fuse_opt_add_arg(&args, arg_buffer);

    result = fuse_main(args.argc, args.argv, &partfs_operations, &context);
    controlled_exit(&context, result);

    return result;
}
